rclcpp_action/include/rclcpp_action/client.hpp

```diff
@@ -562,6 +562,18 @@ class Client : public ClientBase
     return async_cancel(cancel_request, cancel_callback);
   }
 
+  void drop_goal_handle(typename GoalHandle::SharedPtr goal_handle)
```

Comment:
besides https://github.com/ros2/rclcpp/issues/2265#issuecomment-1685839599, i really do not understand why user needs this API to drop the goal handle. user application can drop the `std::shared_future<typename GoalHandle::SharedPtr>` anytime, that tells the system that user is not interested in that anymore.

> Now the handle is dropped, and the object the callbacks are pointing too is invalid.

Why? in that case, `if (!goal_handle) {` comes true? and which is also expected behavior that user does not need it anymore? I am not sure if that gets to the `calling invalid callback`.

```c++
    typename GoalHandle::SharedPtr goal_handle = goal_handles_[goal_id].lock();
    // Forget about the goal if there are no more user references
    if (!goal_handle) {
      RCLCPP_DEBUG(
        this->get_logger(),
        "Dropping weak reference to goal handle during feedback callback");
      goal_handles_.erase(goal_id);
      return;
    }
    auto feedback = std::make_shared<Feedback>();
    *feedback = feedback_message->feedback;
    goal_handle->call_feedback_callback(goal_handle, feedback);
```

Reply:
refer to https://github.com/ros2/rclcpp/issues/2265#issuecomment-1685839599

> We jump back to Thread1. This thread has a internal copy of the shared pointer to the handle, and will call the callback.

as mentioned above, at this moment, `rclcpp` has the shared pointer to the managed object `goal_handle` so this will be valid.

https://github.com/ros2/rclcpp/blob/fd229d72ff498faf6583c1cc1f0f2f131575b2cf/rclcpp_action/include/rclcpp_action/client.hpp#L662

are you saying the `feedback_callback_` will be NULL?

https://github.com/ros2/rclcpp/blob/fd229d72ff498faf6583c1cc1f0f2f131575b2cf/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp#L171-L177

Reply:
```c++
     class Foo {
         GoalHandle::SharedPtr handle;
         void callback();

         //calls async_cancel_goal()
         void cancel_goal();
     };
     std::unique_ptr<Foo> bar = std::make_unique<Foo>();
``` 
handle->feedback_callback points to the method 'callback' of the instance of Foo.
The moment you reset the bar ptr. You can get the race and a call into no longer valid memory.
E.g. the deleted instance of Foo.

Reply:
@fujitatomoya Did my last comment answer you question, do you understand the issue now ?

Reply:
No, not really.

> are you saying the feedback_callback_ will be NULL?

so the answer is yes with your sample?

or some of `Foo` object will be destroyed, and after that using that object in the feedback callback would be the problem? but that because user application intentionally frees the memory?

i would like to hear 2nd opinion here.

Reply:
> are you saying the feedback_callback_ will be NULL?

No, feedback_callback_ contains a value, pointing to &Foo::callback 

> or some of Foo object will be destroyed, and after that using that object in the feedback callback would be the problem? 

Yes

> but that because user application intentionally frees the memory?

Yes, but as you can see, we are in a valid code path, we candled the goal, and afterwards destroyed the object.
And this is exactly the reason, why we need the drop_goal_handle function.

@fujitatomoya can you ping someone for this ?
@kohrt I don't know how to explain it any better, can you give it a try ?

Reply:
@jmachowinski @fujitatomoya 

I understand this change.   

At first, I describe the first problem.
During checking code of action client, I found the current implementation has the problem. 

According to the below declaration, action client only want to keep weak pointer to GoalHandle. (That is, users are responsible for holding the returned shared pointer of GoalHandle)
https://github.com/ros2/rclcpp/blob/ba175922d3008704c67b10fe82a2f7c0cbee2be4/rclcpp_action/include/rclcpp_action/client.hpp#L757

But actually, action client holds the shared pointer of GoalHandle at here
https://github.com/ros2/rclcpp/blob/ba175922d3008704c67b10fe82a2f7c0cbee2be4/rclcpp_action/include/rclcpp_action/client.hpp#L702-L733

The codes should be changed to use weak pointer of GoalHandle here.    

Current example code have the problem (Not hold returned GoalHandle). This is mentioned by your discussion (https://github.com/ros2/rclcpp/issues/2265#issuecomment-1668389518). I check test codes, returned GoalHandle are all held. So I believe the original design intended for users to hold the returned GoalHandle.

The above modification can partially address issue #2265, but it cannot completely solve the problem.    

The cause is

```c++
     class Foo {
         GoalHandle::SharedPtr handle;
         void callback();  // set to call_feedback_callback of GoalHanle
     };
     std::unique_ptr<Foo> bar = std::make_unique<Foo>();
     bar.reset();
```

While calling handle_feedback_message, the object bar is destroyed after `goal_handle = goal_handles_[goal_id].lock();`
Calling goal_handle->call_feedback_callback() still lead to access destroyed resource.

```c++
  void
  handle_feedback_message(std::shared_ptr<void> message) override
  {
    std::lock_guard<std::mutex> guard(goal_handles_mutex_);
    ....
    typename GoalHandle::SharedPtr goal_handle = goal_handles_[goal_id].lock();
    // Forget about the goal if there are no more user references
    if (!goal_handle) {
      RCLCPP_DEBUG(
        this->get_logger(),
        "Dropping weak reference to goal handle during feedback callback");
      goal_handles_.erase(goal_id);
      return;
    }
    auto feedback = std::make_shared<Feedback>();
    *feedback = feedback_message->feedback;
    goal_handle->call_feedback_callback(goal_handle, feedback);
  }
```

So add a new function `drop_goal_handle()` to directly remove GoalHanle in goal_handles_can. This operation avoid this special case.  `drop_goal_handle()` is called in destructor function of Foo.

Reply:
@Barry-Xu-2018 thanks for the explanation, i was asking that via https://github.com/ros2/rclcpp/pull/2281/files#r1304621676. but a bit confused...

now it is clear for me.

Reply:
So I took a look at this issue as well.  Thanks to the excellent explanations by @jmachowinski and @Barry-Xu-2018 , I was able to understand the issue.

While this solution will fix the problem, I think we should explore the solution space further before we commit to this one.  The problem in my mind with this solution is that it adds to the already complex rclcpp_action API, and it will be extremely easy for users to miss calling this in the appropriate places.  In the latter case, users will likely run just fine for a long time without it, until they change to a multi-threaded executor or happen to hit the race.  So I think doing something more "automatic" might be nice, if we can get there.

So with that said, here are some ideas (just throwing them out there, I'm not 100% certain of the feasibility of each of them):

1.  This solution, which has the downsides I've mentioned above but definitely fixes the issue.
2.  Change the parameter type of the feedback, result, and response callbacks from a `GoalHandle::SharedPtr` to something different.  This "something different" would act like a shared/weak pointer, but with an important difference; it would hold a (blocking) mutex while altering the reference count.  That would ensure that the user code could *not* drop the reference while another thread was in the middle of using it.  This has some downsides in that it is easier for the user to create a deadlock, but it is a much more automatic solution that doesn't require a new API.
3.  Since the fundamental problem here is that user code is destroying associated data while an object is still active, we could provide some kind of callback to the user code telling it when we are about to use it.  Then the user code could take appropriate steps.  I'm not a huge fan of this solution, as it has most of the downsides of 1, while also being hard to use.  I just threw it in here as something I thought about while mulling this over.

Thoughts on any of the above, or other thoughts on how we could make this more automatic?

Reply:
@jmachowinski i am with @clalancette on this, probably we can explore the ideas a bit.

the followings are just notes, i will share if i can find any better solution.

> This has some downsides in that it is easier for the user to create a deadlock

yeah, this has reported before, that is the reason we release the mutex lock during the callback...

Reply:
The only 'automatic' solution I can think of would be to pass a weak ptr to the client to the goal handle. In the destructor the goal handle can than try to aquire the client and deregister itself using the mutex. Additional we can make the mutex recursive to try to prevent deadlocks. @clalancette @fujitatomoya what do you think about this approach ?

Reply:
> The only 'automatic' solution I can think of would be to pass a weak ptr to the client to the goal handle. In the destructor the goal handle can than try to aquire the client and deregister itself using the mutex. Additional we can make the mutex recursive to try to prevent deadlocks. @clalancette @fujitatomoya what do you think about this approach ?

I think it is certainly worth a try to see if that solves the issue.

Reply:
Implemented a first version, feedback would be welcome.

Reply:
@jmachowinski thanks for the fix. i think this works.

CC: @clalancette @iuhilnehc-ynos

Reply:
@clalancette @iuhilnehc-ynos ping

Reply:
From the point of my view, it seems that the issue is not being fixed.

The destructor of `ClientGoalHandle` can't be called if the `ClientGoalHandle` is acquired(locked) in the callback.
which means if the user instance is destroyed at the moment, the callback on user side can be called, and then it could be crashed.
(Notice that the reference count of shared_ptr<ClientGoalHandle> with value(2) can not make the ClientGoalHandle destructor called when user destroy the instance which has the `shared_ptr<ClientGoalHandle>`.)

Reply:
@iuhilnehc-ynos  You are right, this won't work. 

Anyone an idea of how to solve this, or is the drop_goal_handle the way to go ?

Reply:
What about taking the handle_mutex_ in the destructor of the ClientGoalHandle and use code like this :
```c++
      typename GoalHandle::SharedPtr goal_handle = goal_handles_[goal_id].lock();
      if (!goal_handle) {
        return;
      }
      // as long as we have this mutex the destructor of the handle can not run through
      // therefore using a raw ptr will be safe here.
      goal_handle->handle_mutex_.lock();
      
      auto *raw_goal_handle_ptr = goal_handle.get();
      
      // from now in any user code dropping the share_ptr, will trigger the destructor of the ClientGoalHandle
      goal_handle.reset();
      if (!goal_handles_[goal_id].use_count() == 0) {
      {
        // last share_ptr pointing to the handle was deleted while we aquired the raw_goal_handle_ptr
        // note, the mutex was deleted, and we therefore do not call unlock on it.
        return;
      }

      // do something
      
      
      raw_goal_handle_ptr->handle_mutex_.unlock();
```

Reply:
Hm, I had a second thought about the solution above, the problem is, that one would need to reinstanciate the shared_ptr for passing it to the callbacl, and the we are back to the race condition. I'm out of ideas....

Reply:
@jmachowinski thanks for consideration.

actually i did not see https://github.com/ros2/rclcpp/pull/2281/files#r1338111143 racy condition, but that could happen. (@iuhilnehc-ynos thanks.)

to be honest, i am not sure either way we go. probably we can keep the issue open until we come up  with feasible solution? any thoughts?

Reply:
As this is a fix for a real world problem, that is affecting at least one party (us ;-) ) I would vote for picking a solution within a bounded time frame and then merging it.